### PR TITLE
Tests - Add test cases for post-process modules in GPT-like models to numerical tests

### PR DESCRIPTION
**Pull request by yzygitzh**
Add test cases for post-process modules in GPT-like models to numerical tests. Including
- Logits.
- Loss.

### CODE REVIEW

File: tests/numerical_tests/modules/test_loss.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    ])
+    @pytest.mark.parametrize('steps', [10])
+    def test_vocab_parallel_cross_entropy(self, inputs_kv, config_kv, loss_kv, steps, request):
+        config = TransformerConfig(**config_kv)
+
+        if config.bf16:
+            dtype = torch.bfloat16
+        else:
+            dtype = torch.float32
```
Comment (cp5555) [inline tests/numerical_tests/modules/test_loss.py:39]: Should we support all types? For example, FP16, FP8, or future FP6, FP4. What's the consideration to limit the type?

File: tests/numerical_tests/modules/test_loss.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            'num_layers': 1,
+            'num_attention_heads': 1,
+        },
+    ])
+    @pytest.mark.parametrize('loss_kv', [
+        {
+            'vocab_size': 200064,
+        }
+    ])
+    @pytest.mark.parametrize('steps', [10])
```
Comment (cp5555) [inline tests/numerical_tests/modules/test_loss.py:32]: Why not put all setting in the config? Then, it can be the default setting?

File: tests/numerical_tests/modules/test_logits.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            'hidden_size': 5120,
+            'num_attention_heads': 1,
+        },
+    ])
+    @pytest.mark.parametrize('logits_kv', [
+        {
+            'vocab_size': 200064,
+        }
+    ])
+    @pytest.mark.parametrize('steps', [10])
```
Comment (cp5555) [inline tests/numerical_tests/modules/test_logits.py:34]: Why not load it from config?

File: tests/numerical_tests/modules/test_logits.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    def test_logits(self, inputs_kv, config_kv, logits_kv, steps, request):
+        config = TransformerConfig(**config_kv)
+
+        if config.bf16:
+            dtype = torch.bfloat16
+        else:
+            dtype = torch.float32
```
Comment (cp5555) [inline tests/numerical_tests/modules/test_logits.py:41]: The similar question about type: what's the limitation for just 2 types?

File: tests/numerical_tests/modules/test_logits.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+import pytest
+import torch
+
+from megatron.core import tensor_parallel
+from megatron.core.transformer import TransformerConfig
+from tests.numerical_tests.modules.test_module import TestModule
+
+
+class TestLogits(TestModule):
```
Comment (cp5555) [inline tests/numerical_tests/modules/test_logits.py:9]: Do you test optimizer? logits are not the same as an optimizer. They serve different roles in deep learning:

Logits:
Logits refer to the raw scores or unnormalized log probabilities output by a neural network's final layer before applying an activation function like softmax. They are used to compute the loss function in tasks such as classification. For example, in classification tasks, logits are transformed into probabilities using the softmax function, which are then used to calculate the loss (e.g., cross-entropy loss) against the true labels.

Optimizer:
An optimizer is an algorithm used to update the weights of a model to minimize the loss function. Common optimizers include SGD (Stochastic Gradient Descent), Adam, and RMSprop.
Optimizers adjust the model parameters by calculating gradients through backpropagation and using these gradients to improve the model's performance on the training data.

Therefore, logits and optimizer are distinct concepts in deep learning, with logits relating to the model's output and optimizer relating to the training process.

**Comment by cp5555** [comment]
Change the title "Tests - Add test cases for loss functions and optimizers in transformer-based models."

**Review by abuccts — APPROVED** [review]
